src/Interfaces/AbstractPolytope.jl

```diff
@@ -96,6 +96,39 @@ function isempty(P::AbstractPolytope)::Bool
     return isempty(vertices_list(P))
 end
 
+"""
+    isuniversal(P::AbstractPolytope{N}, [witness]::Bool=false
+               )::Union{Bool, Tuple{Bool, Vector{N}}} where {N<:Real}
+
+Check whether a polyhedron is universal.
+
+### Input
+
+- `P`       -- polyhedron
+- `witness` -- (optional, default: `false`) compute a witness if activated
+
+### Output
+
+* If `witness` option is deactivated: `false`
+* If `witness` option is activated: `(false, v)` where ``v ∉ P``
+
+### Algorithm
+
+A witness is produced using `isuniversal(H)` where `H` is the first linear
+constraint of `P`.
+"""
+function isuniversal(P::AbstractPolytope{N}, witness::Bool=false
+                    )::Union{Bool, Tuple{Bool, Vector{N}}} where {N<:Real}
+    if witness
+        constraints = constraints_list(P)
+        @assert !isempty(constraints) "polytopes without constraints are not " *
```

Comment:
Why "not allowed"? The following command works:

```julia
julia> using LazySets

julia> HPolytope()
HPolytope{Float64}(HalfSpace{Float64,VN} where VN<:AbstractArray{Float64,1}[])
```

Reply:
Yes, the command is allowed, but the instance should only be used after adding bounding constraints. The instance with no constraints is contradictory and erroneous. If we return `true` here (because the polytope does not have any constraints), this adds even more contradictions.
```julia
julia> isbounded(HPolytope())
true

julia> constrained_dimensions(HPolytope())
0-element Array{Int64,1}

julia> ρ([1.], HPolytope())
ERROR: the support function in direction [1.0] is undefined because
the polytope is unbounded

julia> an_element(HPolytope())
ERROR: ArgumentError: An index is out of bound.

julia> [0.] ∈ HPolytope()
ERROR: AssertionError: a 1-dimensional point cannot be an element
of a -1-dimensional set
```

Reply:
I think you are thinking it too much. The method `isuniversal` for a `HPolytope()` should return `true`, because it is universal, the same way as you added for `HPolyhedron()`. There is nothing more on that.

Reply:
`isuniversal(X) && isbounded(X)` should never hold. This is contradictory. For `AbstractPolytope`, the type implies boundedness and hence non-universality.

Reply:
The title of this PR is "isuniversal methods for bounded sets", but i didn't comment it was contradictory / meaningless, although that thought came into my mind as you point out (it should never hold). But here, if `isuniversal(HPolyhedron())` is `true`, there is no good reason to no let `isuniversal(HPolytope())` return `true` as well.. 

> the type implies boundedness and hence non-universality.

Yes, and in that case, `isuniversal` should *not* make checks with the value, and instead return `false` for any `HPolytope`. But that is not what you have written..

Reply:
> Yes, and in that case, isuniversal should not make checks with the value, and instead return false for any HPolytope. But that is not what you have written

The method returns `false` unless we produce a witness. For that I use the first constraint. This constraint must exist because otherwise the object does not represent a polytope, which the type implies.

Reply:
I see two options:
1. We let `isuniversal(::AbstractPolytope) == false` independent of the concrete object (this is implemented here). Obviously, witness production should not alter that result, so the result should either be `(false, w)` for witness `w` or an error. For `HPolytope`s without any constraints I do not see how to compute `w` - we do not even know the dimension of the set.
2. We let `isuniversal(HPolytope()) == true`, but then we should remove `HPolytope` because obviously there would not be a semantic difference to `HPolyhedron` anymore.

Reply:
Eventually we can add `isboundedtype` to refer to the type and `isbounded` to refer to the object. I find that better than the the current implementation, which is

```julia
function isbounded(P::HPolytope, use_type_assumption::Bool=true)::Bool
    if use_type_assumption
        return true
    end
    return isbounded(HPolyhedron(P.constraints))
end
```
Regarding the universal case, i think that the implementation should be:

```julia
function isuniversal(P::HPolytope, use_type_assumption::Bool=true)::Bool
    if use_type_assumption
        return true
    end
    return isuniversal(HPolyhedron(P.constraints))
end
```
In particular, the assertion should be removed, because it is a value, and because it is wrong: polytopes without constraints *are* allowed as one can do `HPolytope()`.

Reply:
My concern is **only about witness production**. I only want to throw an error in that case. Your proposal does *not* answer that question. What should `isuniversal(HPolytope(), true; use_type_assumption=true)` return?

```julia
function isuniversal(P::HPolytope, witness::Bool=false; use_type_assumption::Bool=true)::Bool
    if use_type_assumption
        if witness
            # ???
        else
            return false
        end
    end
    ...
end
```

> In particular, the assertion should be removed, because it is a value, and because it is wrong: polytopes without constraints are allowed as one can do `HPolytope()`.

Only because you can do `HPolytope()` does not imply that you are allowed to use the result for anything. I personally do not like that you can do that because it clearly violates type assumptions, and I would argue that we only have this option for historical reasons. I would be happy to remove this constructor and rather require that `HPolytope`s must be constructed from a list of bounding constraints. We have `HPolyhedron` for the general case now and you can always `convert` between the types efficiently. But that discussion is going beyond this issue.

Reply:
The following compromise might work for both of us:

* Use a similar implementation for `AbstractPolytope` as for `AbstractPolyhedron` - but with `false` instead of `true`.

```julia
function isuniversal(P::AbstractPolytope{N}, witness::Bool=false
                    )::Union{Bool, Tuple{Bool, Vector{N}}} where {N<:Real}
    return witness ? isuniversal(constraints_list(P)[1], true) : false
end
```

* But for `HPolytope` we override the definition:

```julia
function isuniversal(P::HPolytope{N}, witness::Bool=false
                    )::Union{Bool, Tuple{Bool, Vector{N}}} where {N<:Real}
    constraints = constraints_list(P)
    if isempty(constraints)
        return witness ? (true, N[]) : true
    else
        return isuniversal(constraints[1], true)
    end
end
```

Usually I do not like that overriding an interface definition changes the semantics - it should only provide a more efficient implementation. But in this case the result only differs for `P`s that are (IMO) invalid inputs and for those the output is undefined (for me), so *any* output is good for me and I can also live with inconsistencies in these cases.

What do you think?

Reply:
> What should isuniversal(HPolytope(), true; use_type_assumption=true) return?

It should return `false`, because the type assumption of polytopes is that they are bounded hence non-universal. The witness production option is by definition ill-defined if we only want to use the type assumption, as producing a witness inevitably needs to consider the value.

Reply:
> It should return false

The second argument (`true`) stands for witness production. I disagree that we should not return a witness if a witness was requested.

Reply:
> The witness production option is by definition ill-defined if we only want to use the type assumption, as producing a witness inevitably needs to consider the value.

And this is exactly the case when I throw an error. Then I do not understand why you do not like my solution.

Reply:
Conclusion: `(true, N[])`

Reply:
Ok :+1:  So this will make `isuniversal(HPolytope()) = false` and `isuniversal(HPolytope(), true) = (true, N[])`. I've opened issue https://github.com/JuliaReach/LazySets.jl/issues/1835 to outsource this problem.